feat: refactor Markdown out to @looker/code-editor

noahamac opened the pull request:
b/ https://buganizer.corp.google.com/issues/188459942

## demo
Code block with json decorator
![image](https://user-images.githubusercontent.com/56001784/121990364-835f3800-cd52-11eb-84ed-a9378983c792.png)

Code block without language decorator 
![image](https://user-images.githubusercontent.com/56001784/121990542-ca4d2d80-cd52-11eb-8285-e13900e7e027.png)

Code block and inline support search pattern highlighting
![image](https://user-images.githubusercontent.com/56001784/121990655-f7014500-cd52-11eb-8a69-2749f55de5d7.png)


remaining TODO: 
- Eliminate final `<ReactMarkdown />` usage in ApiExplorer
- Add support for all Prism languages (requirement: all SDK supported langs)
- Decorate more endpoint code blobs (```json) (Please advise)
- more code-editor util tests

noahamac commented on packages/api-explorer/src/components/ExploreType/ExploreProperty.tsx:
```diff
@@ -152,6 +152,7 @@ export const ExplorePropertyDescription: FC<ExplorePropertyProps> = ({
 }) => (
   <>
     {property.description && (
+      // TODO: Replace with @looker/code-editor Markdown
       <ReactMarkdown>{property.description}</ReactMarkdown>
```
TODO 1

noahamac commented on packages/code-editor/src/CodeDisplay/types.ts:
```diff
@@ -0,0 +1,48 @@
+  lineNumbers?: boolean
+}
+
+export type TokenOutputProps = {
+  key?: React.Key
+  style?: any
+  className: string
+  children: string
+  [otherProp: string]: any
+}
+
+export interface LineItemProps {
+  index: number
+  tokenProps: TokenOutputProps
+  pattern: string
+}
```
Copies of unexported Prism types

jkaster replied:
Pity they don't export them. Maybe we should create a PR for Prism?

josephaxisa replied:
any particular reason types are in their own file for the CodeDisplay component? In `api-explorer` and `run-it` types and component implementations coexist in the same file

noahamac commented on packages/code-editor/src/Markdown/utils.ts:
```diff
@@ -0,0 +1,138 @@
+  return { text, language }
+}
+
+/**
+ * Returns a 'qualified markdown' text, a string which contains search pattern match and syntax highlighting qualifiers used by this package
+ * @param content The content to qualify
+ * @param pattern The regex pattern to search
+ * @returns qualified content
+ */
+export const qualifyMarkdownText = (
+  content: string,
+  pattern: string
+): string => {
+  let qualifiedContent
+  if (pattern !== '') {
+    qualifiedContent = addMarkTags(content, pattern)
+  }
+  qualifiedContent = addCodeLanguageTags(qualifiedContent || content)
+  return qualifiedContent
+}
```
ReactMarkdown takes the output of this func as input. 

We add `<mark></mark>` tags to all search matches. 

Then we change code block decorators 
```
\`\`\`json
```
to
```
\`\`\`
<json/>
```
So that the code block handler (CodeDisplay) can check for language tag and use it for syntax highlighting rules and remove it before rendering

jkaster replied:
I'm not sure we need to go this route, but let's discuss.

noahamac commented on packages/code-editor/src/Markdown/utils.ts:
```diff
@@ -0,0 +1,138 @@
+  return content.match(/<json\/>/g) ? 'json' : 'markup'
+}
+
+/**
+ * Removes tags that were applied for syntax highlighting or search pattern matching and returns just the code blob text
+ * @param content - code blob text to render
+ * @returns - rendered code text
+ */
+export const prepareCodeText = (content: string) => {
+  let text = content
+  const language = getCodeLanguageFromTaggedText(text)
+  text = removeCodeLanguageTags(text)
+  text = removeMarkTags(text)
+  text = text.trim()
+  return { text, language }
+}
```
This is the func that removes the `<mark></mark>` tags and `<json/>` tags before rendering. CodeDisplay uses the returned detected language and a searchPattern prop to handle these qualia

jkaster commented on packages/api-explorer/package.json:
```diff
@@ -76,8 +76,6 @@
     "react-router": "^5.1.2",
     "react-router-dom": "^5.1.2",
     "redux": "^4.0.5",
-    "rehype-raw": "^5.1.0",
-    "remark-gfm": "^1.0.0",
```
Love removing dependencies!

jkaster commented on packages/api-explorer/src/components/ExploreType/ExploreProperty.tsx:
```diff
@@ -152,6 +152,7 @@ export const ExplorePropertyDescription: FC<ExplorePropertyProps> = ({
 }) => (
   <>
     {property.description && (
+      // TODO: Replace with @looker/code-editor Markdown
```
We may want this to be a compact version of the `CodeDisplay` component, too

jkaster commented on packages/code-editor/src/Markdown/Markdown.spec.tsx:
```diff
@@ -28,7 +28,7 @@ import React from 'react'
 import { renderWithTheme } from '@looker/components-test-utils'
 
-import { renderWithSearch } from '../../test-utils'
+import { renderWithSearch } from '../../../api-explorer/src/test-utils'
```
yup. imports get a bit ugly here. Not your fault, just something I noticed when I first started playing with moving this component.

jkaster commented on packages/code-editor/src/Markdown/Markdown.spec.tsx:
````diff
@@ -92,6 +94,18 @@ describe('Markdown', () => {
     ).toBeInTheDocument()
   })
 
+  test('it renders syntax highlighted code blocks', () => {
+    const code = '```json\n{\n"model":"thelook"}```'
+    renderWithTheme(<Markdown source={code} />)
+    expect(screen.getByText('"model"')).toHaveClass('property')
+  })
+
+  test('it renders search matches in syntax highlighted code blocks', () => {
+    const code = '```json\n{\n"model":"thelook"}```'
+    renderWithTheme(<Markdown source={code} pattern={'model'} />)
+    expect(screen.getByText('"model"')).toHaveClass('match')
+  })
````
Nice

jkaster commented on packages/code-editor/src/Markdown/utils.spec.ts:
```diff
@@ -23,32 +23,32 @@
+        'query'
       )
       expect(result).toEqual(
         'Here is some text [<mark>Query</mark>](#!/methods/<mark>Query</mark>/create_<mark>query</mark>)'
```
Ironically, this test was verifying bad markings. It should be:

'Here is some text [<mark>Query</mark>](#!/methods/Query/create_query)'

Although perhaps we currently fix that up in the link processor

noahamac replied:
Interesting... nice catch. After looking into it, it looks like ReactMarkdown properly negotiates the `<mark></mark>` tags with the `<a></a>` as the mark tags are only applied to the anchor's text.

jkaster commented on packages/code-editor/src/Markdown/utils.ts:
```diff
@@ -0,0 +1,138 @@
+ * @param markedText - the markdown text input that contains <mark /> tags
+ */
+const removeMarkTags = (markedText: string) => {
+  return markedText.replace(/<mark>/g, '').replace(/<\/mark>/g, '')
```
I think we can use this and it should run faster
```suggestion
  return markedText.replace(/<\/?mark>/g, '')
```

jkaster commented on packages/code-editor/src/Markdown/utils.ts:
````diff
@@ -0,0 +1,138 @@
+  let languageTaggedContent
+  try {
+    const replacement = () => '```\n<json/>'
+    const searchPattern = '```json'
````
I think we can make this expression a capture group so we can retrieve any language
```suggestion
    const searchPattern = /```(.*)$/gm
```

then your replacement would take the captured group and use that

jkaster commented on packages/code-editor/src/Markdown/utils.ts:
```diff
@@ -0,0 +1,138 @@
+  let untaggedContent
+  try {
+    const replacement = () => ''
+    const searchPattern = '<json/>'
```
Here we'd need a pattern that lists all our supported languages

jkaster commented on packages/code-editor/src/Markdown/utils.ts:
```diff
@@ -0,0 +1,138 @@
+ * @returns
+ */
+const getCodeLanguageFromTaggedText = (content: string): string => {
+  return content.match(/<json\/>/g) ? 'json' : 'markup'
```
I think we'll need to have a conversation before I can review this `json` code further.

jkaster reviewed (changes requested):
This is a great step forward.  I have some suggestions and some questions. Let's meet.

noahamac commented on packages/code-editor/src/Markdown/Markdown.tsx:
```diff
@@ -93,10 +89,15 @@ export const Markdown: FC<MarkdownProps> = ({
-    if (inline) {
-      return <MDCode>{children}</MDCode>
+    const { text, language } = prepareCodeText(children?.toString() || '')
+    const codeProps = {
+      language: language,
+      code: text,
+      pattern: pattern,
+      lineNumbers: false,
+      inline: inline,
     }
-    return <MDCodeBlockWrapper value={children} />
+    return <CodeDisplay {...codeProps} />
```
here

bryans99 commented on yarn.lock:
```diff
@@ -14600,10 +14600,10 @@ trim-trailing-lines@^1.0.0:
-  integrity sha512-3JVP2YVqITUisXblCDq/Bi4P9457G/sdEamInkyvCsjbTcXLXIiG7XCb4kGMFWh6JGXesS3TKxOPtrncN/xe8w==
+trim@0.0.1:
+  version "0.0.1"
+  resolved "https://registry.yarnpkg.com/trim/-/trim-0.0.1.tgz#5858547f6b290757ee95cccc666fb50084c460dd"
```
We seem to have lost a resolution here. We were using 1.0.1 but now we are back to 0.0.1. 1.0.1 is the latest

bryans99 reviewed (commented):
Looking good. A couple of small comments. Still need to test it out though

noahamac commented on packages/code-editor/src/CodeDisplay/CodeDisplay.tsx:
```diff
@@ -80,39 +58,49 @@ const LineContent = styled(Span)`
 /**
  * Provides a view-only syntax highlighter for all supported SDK languages.
- * TODO: LookML syntax highlighting
+ * @param language - highlighting language
```
If the docstring is around the prop type def it isn't included in vsCode's docstring tooltip... Maybe there is a way around this?

noahamac replied:
Or do you not use the tooltip?
![image](https://user-images.githubusercontent.com/56001784/122493668-f3f99500-cf9c-11eb-9add-5067debc0d8f.png)

jkaster replied:
It's a challenge if different IDEs use comment docs differently but at this point we're probably best favoring the behavior of VS Code, then IntelliJ for purely selfish and popularity reasons.

noahamac commented:
Added support for markdown code block decorators. For example: 
![Untitled presentation (16)](https://user-images.githubusercontent.com/56001784/122621664-08946680-d04b-11eb-8d47-579bf5374468.png)
![Untitled presentation (17)](https://user-images.githubusercontent.com/56001784/122621696-28c42580-d04b-11eb-9f54-9c4f1949f26e.png)

The default code block highlighting is `markup`. Added support for Ruby, Kotlin, C#, and Swift. 
![Untitled presentation (18)](https://user-images.githubusercontent.com/56001784/122621702-3083ca00-d04b-11eb-8eb9-b62dcee6aa7c.png)

Also was able to utilize this new packaged Markdown component in the Explore Tree and fix the spacing bug there. 

![Untitled presentation (19)](https://user-images.githubusercontent.com/56001784/122621732-472a2100-d04b-11eb-8f1c-b16abb39265c.png)

noahamac commented:
Some questions:
1) What is the `sdk-codegen` style guide on docstrings? For FCs, should the docstring go above the component, should the props be doc'd in the type, etc?
2) How robust of a readme should go in the code-editor dir?

noahamac commented on packages/code-editor/src/Markdown/utils.spec.ts:
```diff
@@ -23,36 +23,26 @@
-        'Here is some text [Query](#!/methods/Query/create_query)'
-      )
-      expect(result).toEqual(
-        'Here is some text [<mark>Query</mark>](#!/methods/<mark>Query</mark>/create_<mark>query</mark>)'
```
Removing test because links are handled elsewhere

jkaster commented on packages/code-editor/src/Markdown/utils.ts:
````diff
@@ -0,0 +1,141 @@
+  try {
+    const searchPattern = /```([A-Za-z]+)$/gm
+    const match = searchPattern.exec(content)
+    const language = match && match[1]
````
what happens if no language is specified? I'd like to see tests for these functions.

noahamac replied:
If no language is specified, the code is styled as `markup`. See `prepareCodeText` and `getCodeLanguageFromTaggedText`

jkaster reviewed (changes requested):
The README.md is empty?

I'd like to see tests for the language handling functions in Markdown/utils

josephaxisa commented on packages/code-editor/package.json:
```diff
@@ -48,7 +52,7 @@
     "react": "^16.13.1",
     "react-dom": "^16.13.1",
     "react-is": "^16.13.1",
-    "styled-components": "^5.2.1",
```
might be worth bumping @types/styled-components as well

noahamac replied:
We're up to date with @looker/components. I tried bumping the @types but it threw errors on how we extend Looker components

josephaxisa commented on packages/code-editor/src/utils/utils.ts:
```diff
@@ -24,16 +24,21 @@
+  return (
+    Object.keys(Prism.languages).includes(languageName) ||
+    extraHighlightingEngines.includes(languageName)
+  )
```
minor nit: for consistency this could be converted into an arrow function

josephaxisa commented on packages/code-editor/src/utils/utils.ts:
```diff
@@ -43,27 +48,36 @@ function instanceOfPrismLanguage(object: any): boolean {
-  if (transparent) {
-    prismTheme.plain.backgroundColor = 'none'
-    prismTheme.plain.padding = '0px'
+export function getOverridenTheme(
```
ditto

josephaxisa commented on packages/code-editor/src/Markdown/utils.spec.ts:
```diff
@@ -0,0 +1,133 @@
+      const result = qualifyMarkdownText(str, '')
+      expect(result).toEqual(str)
+    })
+    test('it returns original string when code block syntax is unspecified', () => {
```
I find it's so much easier for me to read tests when there's some white space between one test and another, and you did it elsewhere hence I thought I'd mention it. Feel free to ignore these
```suggestion

    test('it returns original string when code block syntax is unspecified', () => {
```

josephaxisa commented on packages/code-editor/src/Markdown/utils.spec.ts:
```diff
@@ -0,0 +1,133 @@
+      const result = qualifyMarkdownText(str, '')
+      expect(result).toEqual(str)
+    })
+    test('it returns json syntax tag for json specified syntax', () => {
```
```suggestion

    test('it returns json syntax tag for json specified syntax', () => {
```

josephaxisa commented on packages/code-editor/src/Markdown/utils.spec.ts:
```diff
@@ -0,0 +1,133 @@
+      const result = getCodeLanguageFromTaggedText(str)
+      expect(result).toEqual('markup')
+    })
+    test('it returns `json` for json tagged code text', () => {
```
```suggestion

    test('it returns `json` for json tagged code text', () => {
```

josephaxisa commented on packages/code-editor/src/Markdown/utils.spec.ts:
```diff
@@ -0,0 +1,133 @@
+      const result = getCodeLanguageFromTaggedText(str)
+      expect(result).toEqual('json')
+    })
+    test('it removes json tag for code text display', () => {
```
```suggestion

    test('it removes json tag for code text display', () => {
```

josephaxisa commented on packages/code-editor/src/Markdown/utils.spec.ts:
```diff
@@ -0,0 +1,133 @@
+      const result = removeCodeLanguageTags(str)
+      expect(result).toEqual(toMdString(codeText))
+    })
+    test('it only removes syntax highlighting tags', () => {
```
```suggestion

    test('it only removes syntax highlighting tags', () => {
```

josephaxisa commented on packages/code-editor/src/Markdown/utils.spec.ts:
```diff
@@ -0,0 +1,133 @@
+      const result = removeCodeLanguageTags(str)
+      expect(result).toEqual(str)
+    })
+    test('it adds syntax highlighting tags', () => {
```
```suggestion

    test('it adds syntax highlighting tags', () => {
```

josephaxisa reviewed (commented):
Looks good to me. Did you intend to leave the code editor's readme empty?

jkaster reviewed (changes requested):
Code looks good. I don't see a README.md any more, and we need one.

jkaster reviewed (approved):
That readme looks great!
